### PR TITLE
Set upstream node for branch node

### PR DESCRIPTION
**Pull request by pingsutw**
Signed-off-by: Kevin Su <pingsutw@apache.org>

# TL;DR
We didn't set the upstream node for the branch node. Therefore, the DAG of following example is wrong,
```python
from flytekit import task, workflow, conditional


@task
def square(n: float) -> float:
    return n * n


@task
def double(n: float) -> float:
    return 2 * n


@task
def t1() -> float:
    return 3.0


@workflow
def wf(my_input: float) -> float:                                                                            
    val = t1()
    return (
        conditional("fractions")
            .if_((my_input >= 0.1) & (my_input <= 1.0))
            .then(double(n=val))
            .else_()
            .then(square(n=my_input))
    )

```
Before:
image

After:
image

For more detail, check https://flyte-org.slack.com/archives/C03CY9S9MFE/p1651262585261169 and https://flyte-org.slack.com/archives/C03CY9S9MFE/p1651307245028089

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by wild-endeavor** [comment]
using your example, are these equivalent?
```
@workflow
def wf(my_input: float) -> float:                                                                            
    val = t1()
    return (
        conditional("fractions")
            .if_((my_input >= 0.1) & (my_input <= 1.0))
            .then(double(n=val))
            .else_()
            .then(square(n=my_input))
    )
```

```
@workflow
def wf(my_input: float) -> float:                                                                            
    return (
        conditional("fractions")
            .if_((my_input >= 0.1) & (my_input <= 1.0))
            .then(double(n=t1()))
            .else_()
            .then(square(n=my_input))
    )
```
@kumare3 @pingsutw

**Comment by pingsutw** [comment]
yeah, but if IIRC, the second one doesn't work, right?

**Comment by wild-endeavor** [comment]
should they be equivalent?

**Comment by pingsutw** [comment]
cc @wild-endeavor 
sorry, I think I was wrong. They are not quite equivalent
In the first one, we'll run `t1()` no matter the condition is true or false.
In the second one, we'll run `t1()` only when the condition is true. Therefore, we should make t1 become part of the branch node as well. We don't support it for now, right? 

which means we can't have a workflow like
 ```
                 if  ->  t1 -> t2 
start ->                               -> end
                  else  -> t3
```
For now, we can only have exactly one node in `if` block and `else` block.

**Comment by wild-endeavor** [comment]
yes @pingsutw agreed.  @kumare3 @enghabu what should we do about this?

**Comment by EngHabu** [comment]
I don't think the problem with the example above is the graph... it's in the execution path...
If you look at the generated DAG, it has an upstream dependency between `n1-n0` (the first branch) to `n0` which should cause that node to wait until data from n0 become available before proceeding.

image

Somewhere between [this](https://github.com/flyteorg/flytepropeller/blob/master/pkg/controller/nodes/branch/handler.go) and the Node Handler this breaks....

Please let me know if you need further help investigating this.

**Comment by eapolinario** [comment]
Cleaning stale PRs. Please reopen if you wan to discuss this further.
